content/client-lib-development-guide/features.textile

```diff
@@ -63,15 +63,15 @@ h3(#restclient). RestClient
 ** @(RSC8a)@ "MessagePack":http://msgpack.org/ binary protocol (this is the default for environments having a suitable level or support for binary data)
 ** @(RSC8b)@ JSON text protocol (used when @useBinaryProtocol@ option is false)
 * @(RSC9)@ Uses @Auth@ to establish what authentication scheme to use, how to authenticate, and automatic issuing of tokens when necessary
-* @(RSC10)@ If a REST request responds with 401 with code 40140, then the server is indicating the token is no longer valid. Using the @Auth@ class a token should be automatically issued if possible, otherwise it should result in an error
+* @(RSC10)@ If a REST request responds with 401 HTTP status code and an Ably code value @40140 <= @code < 40150@ indicating that the token has expired or is invalid, then the Auth class should be responsible for reissuing a token and the request should be reattempted, see "RSC14c":#RSC14c and "RSC14d":#RSC14d
 * @(RSC11)@ Requests are sent to the default endpoint @rest.ably.io@. However, if the @host@ option is set, the client will send requests to the specified host. If @environment@ option is configured, the environment name is prefixed to the default host endpoint and the @host@ is set accordingly. For example, if the @environment@ is set to sandbox, then the @host@ endpoint will become @sandbox-rest.ably.io@
 * @(RSC12)@ REST endpoint host is configurable in the Client constructor with the option @restHost@
 * @(RSC13)@ The client library must use the connection and request timeouts specified in the @ClientOptions@, falling back to the defaults described in @ClientOptions@ below
 * @(RSC14)@ Authentication
 ** @(RSC14a)@ Supports basic authentication when an API key is provided with the @key@ option. The API key follows the format @"KEY_NAME:KEY_SECRET"@ so when authenticating using "Basic Auth":https://en.wikipedia.org/wiki/Basic_access_authentication, the key name can be used as the username and the key secret as the password
 ** @(RSC14b)@ Uses token authentication whenever the criteria of "RSA4":#RSA4 are met
-** @(RSC14c)@ When a @token@ or @tokenDetails@ is used to instance the library, and no means to renew the token is provided (either an API key, @authCallback@ or @authUrl@), if the server responds with 40140 (token expired) to a request, the client library should indicate an error and not retry the request
-** @(RSC14d)@ When the client does have a means to renew the token automatically, and the token has expired or the server has responded with 40140 (token expired), then the client should automatically renew the token and resend the request
+** @(RSC14c)@ When a @token@ or @tokenDetails@ is used to instance the library, and no means to renew the token is provided (either an API key, @authCallback@ or @authUrl@), if the server responds with a 401 HTTP status code and an Ably code value @40140 <= @code < 40150@ indicating token is expired or invalid, the client library should indicate an error and not retry the request
```

Comment:
I don't really like the idea that the library now has to remember the previous error code.

An error in the range [40140 ... 40150) is saying that there was a specific issue with that token, and there is every reason to suppose that a new token will not have the same problem. If on the other hand the key was revoked, say, you'd get a 4013xx error and would therefore not attempt to get a new token.

So although I don't like the possibility of a perpetual cycle of issuing tokens, it should not arise provided we return the correct error code.